{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0c33e310",
   "metadata": {},
   "outputs": [],
   "source": [
    "from thymiodirect import Thymio, Connection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ec14ecce",
   "metadata": {},
   "outputs": [],
   "source": [
    "port = Connection.serial_default_port()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9eb142b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24330 is connected\n"
     ]
    }
   ],
   "source": [
    "# Connect to Thymio and store id in variable\n",
    "\n",
    "th = Thymio(serial_port = port, on_connect = lambda id: print (f\"{id} is connected\"))\n",
    "th.connect()\n",
    "id = th.first_node()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d33bb09",
   "metadata": {},
   "outputs": [],
   "source": [
    "## th[id][\"leds.top\"] = [0, 0, 32]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "803d85f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "prox_prev = 0\n",
    "def obs(node_id):\n",
    "    global prox_prev\n",
    "    prox = (th[node_id][\"prox.horizontal\"][5]\n",
    "            - th[node_id][\"prox.horizontal\"][2]) // 10\n",
    "    if prox != prox_prev:\n",
    "        th[node_id][\"motor.left.target\"] = prox\n",
    "        th[node_id][\"motor.right.target\"] = prox\n",
    "        print(prox)\n",
    "        if prox > 5:\n",
    "            th[id][\"leds.top\"] = [0, 32, 0]\n",
    "        elif prox < -5:\n",
    "            th[id][\"leds.top\"] = [32, 32, 0]\n",
    "        elif abs(prox) < 3:\n",
    "            th[id][\"leds.top\"] = [0, 0, 32]\n",
    "        prox_prev = prox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "860f8cbe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-257\n",
      "-243\n",
      "-232\n",
      "0\n",
      "-152\n",
      "-171\n",
      "-254\n",
      "-232\n",
      "0\n",
      "228\n",
      "275\n",
      "314\n",
      "214\n",
      "186\n",
      "146\n",
      "148\n",
      "191\n",
      "232\n",
      "208\n",
      "215\n",
      "185\n",
      "194\n",
      "190\n",
      "180\n",
      "235\n",
      "249\n",
      "237\n",
      "249\n",
      "265\n",
      "266\n",
      "250\n",
      "222\n",
      "0\n",
      "-291\n",
      "-379\n",
      "-342\n",
      "-293\n",
      "-268\n",
      "-251\n",
      "-232\n",
      "-220\n",
      "-213\n",
      "-226\n",
      "-233\n",
      "-213\n",
      "-208\n",
      "-198\n",
      "-196\n",
      "-211\n",
      "-234\n",
      "-259\n",
      "-263\n",
      "-289\n",
      "0\n",
      "128\n",
      "223\n",
      "272\n",
      "254\n",
      "0\n",
      "177\n",
      "281\n",
      "286\n",
      "291\n",
      "302\n",
      "344\n",
      "305\n",
      "276\n",
      "248\n",
      "188\n",
      "193\n",
      "209\n",
      "220\n",
      "186\n",
      "185\n",
      "209\n",
      "159\n",
      "0\n",
      "147\n",
      "164\n",
      "189\n",
      "214\n",
      "186\n",
      "139\n",
      "0\n",
      "125\n",
      "167\n",
      "259\n",
      "261\n",
      "236\n",
      "226\n",
      "210\n",
      "218\n",
      "233\n",
      "208\n",
      "155\n",
      "0\n",
      "125\n",
      "225\n",
      "248\n",
      "253\n",
      "251\n",
      "217\n",
      "208\n",
      "219\n",
      "248\n",
      "284\n",
      "316\n",
      "0\n",
      "-158\n",
      "-250\n",
      "-454\n",
      "-322\n",
      "-427\n",
      "-333\n",
      "-353\n",
      "-328\n",
      "-333\n",
      "-325\n",
      "-309\n",
      "-316\n",
      "-319\n",
      "-302\n",
      "-287\n",
      "-297\n",
      "-302\n",
      "-290\n",
      "-196\n",
      "0\n",
      "252\n",
      "336\n",
      "399\n",
      "319\n",
      "225\n",
      "363\n",
      "286\n",
      "239\n",
      "265\n",
      "0\n",
      "-235\n",
      "-314\n",
      "-426\n",
      "-410\n",
      "-326\n",
      "-342\n",
      "-277\n",
      "-281\n",
      "-342\n",
      "-375\n",
      "-327\n",
      "-333\n",
      "0\n",
      "265\n",
      "310\n",
      "292\n",
      "263\n",
      "0\n",
      "-220\n",
      "-289\n",
      "-411\n",
      "-414\n",
      "-349\n",
      "-297\n",
      "-188\n",
      "0\n",
      "306\n",
      "307\n",
      "304\n",
      "244\n",
      "142\n",
      "0\n",
      "-380\n",
      "-409\n",
      "-396\n",
      "-352\n",
      "-337\n",
      "-292\n",
      "-262\n",
      "-232\n",
      "-204\n",
      "-178\n",
      "-134\n",
      "0\n",
      "317\n",
      "300\n",
      "311\n",
      "346\n",
      "282\n",
      "237\n",
      "184\n",
      "147\n",
      "0\n",
      "-297\n",
      "-375\n",
      "-453\n",
      "-455\n",
      "-402\n",
      "-314\n",
      "-276\n",
      "-236\n",
      "0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-2652' coro=<Connection.handle_message() done, defined at C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py:398> exception=KeyError(254)>\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py\", line 509, in handle_message\n",
      "    self.remote_nodes[source_node].last_msg_time = time.time()\n",
      "KeyError: 254\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-2653' coro=<Connection.handle_message() done, defined at C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py:398> exception=KeyError(32)>\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py\", line 509, in handle_message\n",
      "    self.remote_nodes[source_node].last_msg_time = time.time()\n",
      "KeyError: 32\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-2654' coro=<Connection.handle_message() done, defined at C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py:398> exception=KeyError(0)>\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py\", line 509, in handle_message\n",
      "    self.remote_nodes[source_node].last_msg_time = time.time()\n",
      "KeyError: 0\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-2655' coro=<Connection.handle_message() done, defined at C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py:398> exception=KeyError(22)>\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py\", line 509, in handle_message\n",
      "    self.remote_nodes[source_node].last_msg_time = time.time()\n",
      "KeyError: 22\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-4404' coro=<Connection.handle_message() done, defined at C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py:398> exception=KeyError(256)>\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py\", line 509, in handle_message\n",
      "    self.remote_nodes[source_node].last_msg_time = time.time()\n",
      "KeyError: 256\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-4405' coro=<Connection.handle_message() done, defined at C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py:398> exception=KeyError(32)>\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py\", line 509, in handle_message\n",
      "    self.remote_nodes[source_node].last_msg_time = time.time()\n",
      "KeyError: 32\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-4406' coro=<Connection.handle_message() done, defined at C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py:398> exception=KeyError(0)>\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py\", line 509, in handle_message\n",
      "    self.remote_nodes[source_node].last_msg_time = time.time()\n",
      "KeyError: 0\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-4407' coro=<Connection.handle_message() done, defined at C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py:398> exception=KeyError(23)>\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\timon\\.conda\\envs\\robo-thymio\\lib\\site-packages\\thymiodirect\\connection.py\", line 509, in handle_message\n",
      "    self.remote_nodes[source_node].last_msg_time = time.time()\n",
      "KeyError: 23\n"
     ]
    }
   ],
   "source": [
    "th.set_variable_observer(id, obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "306e0ca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "th.set_variable_observer(id, None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "robo-thymio",
   "language": "python",
   "name": "robo-thymio"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
